pkg/chaosdaemon/ipset_server.go

```diff
@@ -35,59 +43,105 @@ func (s *Server) FlushIpSet(ctx context.Context, req *pb.IpSetRequest) (*empty.E
 	// TODO: lock every ipset when working on it
 
 	set := req.Ipset
-
 	name := set.Name
 
-	{
-		// TODO: Hash and get a stable short name (as ipset name cannot be longer than 31 byte)
-		cmd := withNetNS(ctx, nsPath, "ipset", "create", name+"old", "hash:ip")
+	// If the ipset already exists, the ipset will be renamed to this temp name.
+	tmpName := fmt.Sprintf("%s_old", name)
+
+	// the ipset while existing iptables rules are using them can not be deleted,.
+	// so we creates an temp ipset and swap it with existing one.
+	if err := s.createIPSet(ctx, nsPath, tmpName); err != nil {
+		return nil, err
+	}
+
+	// add ips to the temp ipset
+	if err := s.addIpsToIPSet(ctx, nsPath, tmpName, set.Ips); err != nil {
+		return nil, err
+	}
+
+	// rename the temp ipset with the target name of ipset if the taget ipset not exists,
+	// otherwise swap  them with each other.
+	if err := s.renameIPSet(ctx, nsPath, tmpName, name); err != nil {
+		return nil, err
+	}
+
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) createIPSet(ctx context.Context, nsPath string, name string) error {
+	// ipset name cannot be longer than 31 byte
+	if len(name) > 31 {
+		name = name[:31]
+	}
+
+	cmd := withNetNS(ctx, nsPath, "ipset", "create", name, "hash:ip")
+
+	log.Info("create ipset", "command", cmd.String())
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		output := string(out)
+		if !strings.Contains(output, ipsetExistErr) {
+			log.Error(err, "ipset create error", "command", cmd.String(), "output", output)
+			return err
+		}
+
+		cmd := withNetNS(ctx, nsPath, "ipset", "flush", name)
```

Comment:
Seems it will return error when the ipset is already exist, in another word, the ipset here is a new one. What's the effect of flushing a new ipset.

Reply:
the old ipset will be overwrite.

Reply:
I misunderstand to logic of creation, it looks good.